_posts/2018-01-07-building-serverless-radiator.md

```diff
@@ -0,0 +1,225 @@
+
+```clojure
+(def projects
+  [{:name             "example account"
+    :aws              {:uri     "https://uri-to.com/api-gateway"
+                       :api-key "secret-api-key"}
+    }])
+```
+
+### Supporting different CI providers
+
+The Radiator should of course support also another CI providers like [GitLab](https://about.gitlab.com/). To make this easier I made a generic data structure for pipelines, alarms and metrics. The radiator should therefore first fetch data from endpoints, after that transform them to generic format which is then rendered to Radiator view.
+
+The generic data structure is defined as Clojure Spec as next with example:
+
+```clojure
+(s/def ::name string?)
+(s/def ::pipeline-status #{:success :in-progress :failed :unknown})
+(s/def ::pipeline (s/keys :req-un [::name ::pipeline-status]))
+(s/def ::alarm-status #{:ok :alarm})
+(s/def ::alarm (s/keys :req-un [::name ::alarm-status]))
+(s/def ::metric (s/keys :req-un [::name ::metric-value]))
+
+;; Pipeline example
+{:name "InfraBuild"
+:pipeline-status :in-progress}
+
+;; Alarm example
+{:name "Service-500-alarm"
+:alarm-status :ok}
+
+;; Metric example
+{:name "RDS avg. CPU"
+:metric-value 1.3}
+```
+
+Supporting the GitLab is done simply fetching pipeline data via REST api and after that transforming a pipeline status to generic format.
+
+```clojure
+(defn- transform-pipeline
+  [{:keys [status name]}]
+  (condp = status
+    "success"  {:name name :pipeline-status :success}
+    "failed"   {:name name :pipeline-status :failed}
+    "running"  {:name name :pipeline-status :in-progress}
+    "pending"  {:name name :pipeline-status :in-progress}
+    "canceled" {:name name :pipeline-status :success}
+    "skipped"  {:name name :pipeline-status :success}
+    {:name name :pipeline-status :unknown}))
+```
+
+### Adding visual hints
+
+The radiator should show at a glance what is going on in accounts. In default there is no need to list different pipelines or alarms - only show the number of pipelines etc. so we can be sure that the exposer is working correctly.
+
+![All OK](/img/building-serverless-radiator/all-ok.png)
+
+When the pipeline is running the account name is changed to yellow and the running pipeline is shown:
+
+![Pipeline running](/img/building-serverless-radiator/pipeline-running.png)
+
+Unfortunately there is situations where all gone bad and it should be visually made clear by using red color:
+
+![Alarm!](/img/building-serverless-radiator/alarm.png)
+
+
+###  Installing radiator
+
+The radiator is also installed by using Serverless framework. The framework has support for Java but for easier deployment I made a [Clojure serverless plugin](https://www.npmjs.com/package/serverless-clj-plugin) for better Clojure support. Using the plugin the installation is done simply by running  ``serverless deploy``. The radiator endpoint should of course be protected but instead of API Gateway - token string can be set in the configuration file radiator.config.
```

Comment:
"by using *the* Serverless framework."

En ihan ymmärrä tätä pätkää: "but instead of API Gateway - token string can be set in the configuration file radiator.config."